ca/server.go

```diff
@@ -551,14 +544,41 @@ func (s *Server) signNodeCert(ctx context.Context, node *api.Node) {
 		}).WithError(err).Errorf("failed to parse role")
 		return
 	}
+
 	// Attempt to sign the CSR
 	cert, err := s.securityConfig.RootCA().ParseValidateAndSignCSR(node.Certificate.CSR, node.Certificate.CN, role, s.securityConfig.ClientTLSCreds.Organization())
 	if err != nil {
 		log.G(ctx).WithFields(logrus.Fields{
 			"node.id": node.ID,
 			"method":  "(*Server).signNodeCert",
-		}).WithError(err).Errorf("failed to parse CSR")
-		return
+		}).WithError(err).Errorf("failed to sign CSR")
+		// If this error is due the lack of signer, maybe some other
+		// manager in the future will pick it up. Return without
+		// changing the state of the certificate.
+		if err == ErrNoValidSigner {
+			return
+		}
+		// If the current state is already Failed, no need to change it
+		if node.Certificate.Status.State == api.IssuanceStateFailed {
+			return
+		}
+		// We failed to sign this CSR, change the state to FAILED
+		for {
+			err = s.store.Update(func(tx store.Tx) error {
+				node.Certificate.Status = api.IssuanceStatus{
+					State: api.IssuanceStateFailed,
+					Err:   err.Error(),
+				}
+
+				return store.UpdateNode(tx, node)
+			})
+			if err == nil {
+				return
+			}
+			if err == store.ErrSequenceConflict {
```

Comment:
Is there a way of getting around this `for` loop pattern? `signNodeCert` is too long because of all of these details.